## Goal:  Train a Naive Bayes model to classify future SMS messages as either spam or ham.

Steps:

1.  Convert the words ham and spam to a binary indicator variable(0/1)

2.  Convert the txt to a sparse matrix of TFIDF vectors

3.  Fit a Naive Bayes Classifier

4.  Measure your success using roc_auc_score



In [31]:
import pandas as pd
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cross_validation import train_test_split
from sklearn import naive_bayes
from sklearn.metrics import roc_auc_score

In [32]:
df= pd.read_csv("SMSSpamCollection",sep='\t', names=['spam', 'txt'])

In [33]:
df.head()

spam                                                txt
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [34]:
dummies = pd.get_dummies(df['spam'], prefix='class', drop_first=True)
# Update X to include dummies and drop the main variable
df = pd.concat([df, dummies], axis=1)
df.drop(['spam'], axis=1, inplace=True)

In [35]:
df.head()

txt  class_spam
0  Go until jurong point, crazy.. Available only ...           0
1                      Ok lar... Joking wif u oni...           0
2  Free entry in 2 a wkly comp to win FA Cup fina...           1
3  U dun say so early hor... U c already then say...           0
4  Nah I don't think he goes to usf, he lives aro...           0

In [36]:
#TFIDF Vectorizer, just like before
stopset = set(stopwords.words('english'))
vectorizer = TfidfVectorizer(use_idf=True, lowercase=True, strip_accents='ascii', stop_words=stopset)

In [37]:
#in this case our dependent variable will be liked as 0 (didn't like the movie) or 1 (liked the movie)
y = df.class_spam

In [38]:
#convert df.txt from text to features
X= vectorizer.fit_transform(df.txt)

In [39]:
print (y.shape)
print (X.shape)

(5572,)
(5572, 8587)


In [40]:
#Test Train Split as usual
X_train, X_test,y_train, y_test = train_test_split(X, y, random_state=42)

In [41]:
#we will train a naive_bayes classifier
clf = naive_bayes.MultinomialNB()
clf.fit(X_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [42]:
#We can test our model's accuracy like this:

roc_auc_score(y_test, clf.predict_proba(X_test)[:,1])

0.98589322144123448

## ROC AUC score of .9859 is pretty good, especially considering ease and speed of setting up the Naive Bayes classifier